pcmaudiotrack.go

```diff
@@ -0,0 +1,197 @@
+package lksdk
+
+import (
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	audio "github.com/livekit/mediatransportutil/pkg/audio"
+	opus "github.com/livekit/mediatransportutil/pkg/audio/opus"
+	rtp "github.com/livekit/mediatransportutil/pkg/audio/rtp"
+	protoLogger "github.com/livekit/protocol/logger"
+	"github.com/pion/webrtc/v4"
+	"go.uber.org/atomic"
+)
+
+const (
+	opusSampleRate = 48000
+)
+
+type PCM16ToOpusAudioTrack struct {
+	*webrtc.TrackLocalStaticSample
+
+	opusWriter         audio.WriteCloser[opus.Sample]
+	pcmWriter          audio.WriteCloser[audio.PCM16Sample]
+	resampledPCMWriter audio.WriteCloser[audio.PCM16Sample]
+
+	frameDuration time.Duration
+
+	sampleBuffer []audio.PCM16Sample
+	started      sync.Once
+	ticker       *time.Ticker
+
+	closed atomic.Bool
+	mu     sync.Mutex
+	cond   *sync.Cond
+}
+
+// TODO: Support stereo
+func NewPCM16ToOpusAudioTrack(sampleRate int, frameDuration time.Duration, logger protoLogger.Logger) (*PCM16ToOpusAudioTrack, error) {
+	track, err := webrtc.NewTrackLocalStaticSample(webrtc.RTPCodecCapability{MimeType: webrtc.MimeTypeOpus}, "test", "test")
+	if err != nil {
+		return nil, err
+	}
+
+	opusWriter := audio.FromSampleWriter[opus.Sample](track, opusSampleRate, frameDuration)
+	pcmWriter, err := opus.Encode(opusWriter, 1, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	resampledPCMWriter := pcmWriter
+	if sampleRate != opusSampleRate {
+		resampledPCMWriter = audio.ResampleWriter(pcmWriter, opusSampleRate)
+	}
+
+	t := &PCM16ToOpusAudioTrack{
+		TrackLocalStaticSample: track,
+		opusWriter:             opusWriter,
+		pcmWriter:              pcmWriter,
+		resampledPCMWriter:     resampledPCMWriter,
+		frameDuration:          frameDuration,
+		// TODO: Maybe not the best thing to do, good for a PoC
+		sampleBuffer: make([]audio.PCM16Sample, 0),
+	}
+	t.cond = sync.NewCond(&t.mu)
+	return t, nil
+}
+
+func (t *PCM16ToOpusAudioTrack) WriteSample(sample audio.PCM16Sample) error {
+	if t.closed.Load() {
+		return errors.New("track is closed")
+	}
+
+	var isFirstSample bool
+	var err error
+
+	t.started.Do(func() {
+		isFirstSample = true
+		// write the first sample immediately before starting the ticker
+		err = t.resampledPCMWriter.WriteSample(sample)
+		t.ticker = time.NewTicker(t.frameDuration)
+		go t.processSamples()
+	})
+
+	if isFirstSample {
+		return err
+	}
+
+	t.mu.Lock()
+	t.sampleBuffer = append(t.sampleBuffer, sample)
+	t.cond.Broadcast()
+	t.mu.Unlock()
+	return nil
+}
+
+func (t *PCM16ToOpusAudioTrack) processSamples() {
```

Comment:
unless required outside, we use the pattern of
```
NewTicker()
defer ticker.Stop()
```
in the worker to ensure that it gets cleaned up when the worker exits and no state variable needed for ticker.

Reply:
ticker.Stop is being called on close at the moment, since the loop has to stop when close is called

Reply:
Yeah, for that the `closed` atomic.Bool is there. And the worker goroutine checks that and exits. So, when the goroutine exits, the ticker can be defer stopped. So, it can all be inside the worker.

Reply:
Oh okay yeah